### ISSUE TITLE
Calculate weight and bias gradients

### ISSUE THREAD

**Issue by kevinsuedmersen** [kevinsuedmersen/library-neuralnetworksfromscratch#2]
Use the error at each layer to calculate the gradients w.r.t the weights and biases

Tasks:
- [x] Calculate weight gradients
- [x] Calculate bias gradients

Tests:
- Calculate the gradients using backpropagation, calculate the gradients using a brute force method and then assert that both gradients are equal to each other
    - [x] Implement the brute force method for all weight gradients
- If gradients are unequal, implement the following checks
    - [x] Make sure that both types of learning algorithms get exactly the same data
    - [x] Make sure that both learning algorithms are instantiated with exactly the same parameters. However, make sure not all parameters in the same layer have the same value to avoid that activations in that layer are all equal to 1. Because if all activations in a layer are equal to 1, the Jacobian of the tanh and sigmoid function will contain only zeros
    - [x] When calculating a partial derivative using the brute force method, make sure we always start with the same set of parameters
    - [x] After training the model using both methods, make sure that all gradients have actually changed
    - [x] When doing the initial forward pass with both learning algorithms, make sure that the loss value is exactly the same
    - [x] Make sure that in every test run, the initial cost of both learning algorithms is always the same
- Create a very simple network with one hidden layer, calculate the gradients with pen and paper using the chain rule, calculate them using backprop and assert that both gradients are equal to each other
    - [x] Test computed dendritic potentials and activations
    - [x] Test computed jacobians
    - [x] Test computed errors
    - [x] Test computed weight gradients
    - [x] Test computed bias gradients
